lib/UndoTruncatedSwitchConditionPass.cpp

```diff
@@ -66,30 +131,33 @@ bool UndoTruncatedSwitchConditionPass::runOnModule(Module &M) {
     }
   }
 
+  zombies_.reset();
+  i32_ = Type::getInt32Ty(M.getContext());
+
   for (auto SI : WorkList) {
     auto Cond = SI->getCondition();
 
-    if (auto TI = dyn_cast<TruncInst>(Cond)) {
-      auto Op = TI->getOperand(0);
-      SI->setCondition(Op);
+    auto *widened = ZeroExtend(Cond);
+    SI->setCondition(widened);
+    Changed = true;
 
-      auto OpTy = Op->getType();
+    for (auto Cases : SI->cases()) {
+      // The original value of the case.
+      auto V = Cases.getCaseValue()->getZExtValue();
 
-      for (auto Cases : SI->cases()) {
-        // The original value of the case.
-        auto V = Cases.getCaseValue()->getZExtValue();
+      // A new value for the case with the correct type.
+      auto CI = dyn_cast<ConstantInt>(ConstantInt::get(i32_, V));
 
-        // A new value for the case with the correct type.
-        auto CI = dyn_cast<ConstantInt>(ConstantInt::get(OpTy, V));
-
-        // And we replace the old value.
-        Cases.setValue(CI);
-      }
+      // And we replace the old value.
+      Cases.setValue(CI);
+    }
+  }
 
-      TI->eraseFromParent();
-    } else {
-      Cond->print(errs());
-      llvm_unreachable("Unhandled switch instruction condition!");
+  // Remove the zombies if we can.  We expect to.  They are ordered from
```

Comment:
This is not a perfect ordering for erasure, but it's good enough for the cases I've seen.

Example where it could fail: 
- phis that feed themselves, or mutally.
- the same narrow value feeding multiple switches.